src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -427,95 +357,70 @@ private IProjectTree CreateDependenciesFolder(IProjectTree oldNode)
                     Flags = DependencyTreeFlags.DependenciesRootNodeFlags
                 };
 
-                ApplyProjectTreePropertiesCustomization(null, values);
+                // Allow property providers to perform customization
+                foreach (IProjectTreePropertiesProvider provider in ProjectTreePropertiesProviders.ExtensionValues())
+                {
+                    provider.CalculatePropertyValues(null, values);
+                }
 
                 // Note that all the parameters are specified so we can force this call to an
                 // overload of NewTree available prior to 15.5 versions of CPS. Once a 15.5 build
                 // is publicly available we can move this to an overload with default values for
                 // most of the parameters, and we'll only need to pass the interesting ones.
                 return NewTree(
-                         caption: values.Caption,
-                         filePath: null,
-                         browseObjectProperties: null,
-                         icon: values.Icon,
-                         expandedIcon: values.ExpandedIcon,
-                         visible: true,
-                         flags: values.Flags);
-            }
-            else
-            {
-                return oldNode;
+                    caption: values.Caption,
+                    filePath: null,
+                    browseObjectProperties: null,
+                    icon: values.Icon,
+                    expandedIcon: values.ExpandedIcon,
+                    visible: true,
+                    flags: values.Flags);
             }
         }
 
-        private async Task<IImmutableDictionary<string, IPropertyPagesCatalog>> GetNamedCatalogsAsync(IProjectCatalogSnapshot catalogs)
+        private void OnDependenciesSnapshotChanged(object sender, SnapshotChangedEventArgs e)
```

Comment:
Do we have any tests that will fail if the order changes are processed is not correct?

Reply:
In general it is hard to test such things via unit tests and i was going to suggest you to hold on your refactorings and create first a set of E2E integration tests(Apex) + stress tests, which would shield you from regressions , especially for things like here which can have timing issues.

---
In reply to: [226568463](https://github.com/dotnet/project-system/pull/4143#discussion_r226568463) [](ancestors = 226568463)

Reply:
All you changes look good, however they are basically untested, since manual sanity verification does not count for many scenarios.

---
In reply to: [226698874](https://github.com/dotnet/project-system/pull/4143#discussion_r226698874) [](ancestors = 226698874,226568463)

Reply:
Apex sounds interesting, I'll look into it.

Reply:
Ping me offline, i will send you a TFS link to my integration tests for nesting , they do check the tree and i have a stress test there too. IT is actually pretty handy to have.

Reply:
Thanks. Message sent on Teams.